Add more property based tests for basic bitcoin data structures

Christewart opened the pull request:
This is the second half of #12775. 

In general property based testing is a great testing idiom for making sure invariants that you believe about your code hold true under a range of "valid" values. 

This PR includes basic properties for  

1. [Keys](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/key_properties.cpp)
2. [Blocks](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/block_properties.cpp)
3. [Bloom Filters](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/bloom_properties.cpp)
4. [Merkleblocks](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/merkleblock_properties.cpp)
5. [Scripts](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/script_properties.cpp)
6. [Transactions](https://github.com/bitcoin/bitcoin/blob/adeb5e7b8757c71b25f32d4278fed4218b1d5851/src/test/transaction_properties.cpp)

Currently rapidcheck is not enabled by default on travis -- work has been done to enable in #14171 there appears to be a memory access violation in one of the environments.

Christewart committed:
Add more property based tests for basic bitcoin data structures

MarcoFalke commented on src/test/gen/block_gen.h:
```diff
@@ -0,0 +1,58 @@
+#include "test/gen/crypto_gen.h"
+#include "test/gen/transaction_gen.h"
+#include "uint256.h"
+#include "primitives/block.h"
```
Please use bracket syntax includes ("#include <foo.h>") instead of quote syntax includes:

Christewart committed:
try to make linter happy

fix typo in Makefile.test.include

rebase and redo imports

Rebase onto master, fix compile erros with various API changes in master, address nits

Run clang-format-diff.py

Add missing copyright header

instagibbs commented on src/test/gen/crypto_gen.cpp:
```diff
@@ -14,6 +11,13 @@
     return rc::gen::suchThat(rc::gen::arbitrary<std::vector<CKey>>(), [](const std::vector<CKey>& keys) {
+        //TODO: Investigate why we can only allow 15 keys. Consensus rules
+        // dictate we can up to 20 keys
+        //https://github.com/bitcoin/bitcoin/blob/10bee0dd4f37eb6cb7a0f1d565fa0fecf8109c35/src/script/script.h#L29
```
the wallet has no idea how to make a script using numbers beyond OP_16. Once you do n-of-17 you need a single byte pushing instruction to push the number `17`

Christewart replied:
I think the original reason for this was actually that more than 15 public keys it was too big of a val to push onto the stack (520 byte limit)

practicalswift commented:
Concept ACK

Very nice!

jb55 commented:
Awesome! Concept Ack!

Github's code review won't let me make line comments so I'm going to put
them here:

    +template <>
    +struct Arbitrary<CBlock> {
    +    static Gen<CBlock> arbitrary()
    +    {
    +        return gen::map(gen::nonEmpty<std::vector<CTransactionRef>>(), [](const std::vector<CTransactionRef>& refs) {
    +            CBlock block;
    +            block.vtx = refs;
    +            return block;
    +        });
    +    }

why gen::nonEmpty here?

    /** Returns [0,100) uint256s */
    +Gen<std::vector<uint256>> BetweenZeroAnd100()
    +{
    +    return gen::suchThat<std::vector<uint256>>([](const std::vector<uint256>& hashes) {
    +        return hashes.size() <= 100;
    +    });
    +}

I suspect there's a more efficient way to implement this. Right now it
looks like it would generate many vectors wastefully?

Perhaps something like:

    const auto size = *rc::gen::inRange(0, 100);
    rc::gen::container(size, gen::arbitrary<uint256>())


Skimming through the rest, LGTM. Building now to test locally.

Empact commented:
Ref #12775 which this builds on.

Empact commented on src/test/gen/block_gen.h:
```diff
@@ -0,0 +1,58 @@
+            uint32_t nTime;
+            uint32_t nBits;
+            uint32_t nNonce;
+            std::tie(nVersion, hashPrevBlock, hashMerkleRoot, nTime, nBits, nNonce) = primitives;
```
Seems you could directly `tie` into the `header` fields, to make this more succinct.

Christewart replied:
Not sure what you mean about this

Empact commented:
How about responding to @MarcoFalke's suggestions?
https://github.com/bitcoin/bitcoin/pull/12775#discussion_r216164298
https://github.com/bitcoin/bitcoin/pull/12775#discussion_r216158186

Empact commented on src/test/gen/bloom_gen.cpp:
```diff
@@ -0,0 +1,59 @@
+    });
+}
+
+rc::Gen<unsigned int> Between1And100()
```
nit: looks like this method is defined with the same name here and in `merkleblock_gen.h`, how about:
* applying `static` more liberally?
* renaming one or the other?

Christewart replied:
So basically mark as `static` in merkleblock_gen.h and then include that header file and use it here?

Empact replied:
Well they have different implementations for different return results. Calling this `bloom_gen.cpp` implementation `static` will prevent it from being used outside this file, which would reduce the risk of conflicts and otherwise be appropriate as the method is only used here. I don't think you want to apply `static` in the header, as that will result in the function being compiled into every including translation unit. Naming this to not have a potential conflict would also be positive.

Empact commented on src/test/gen/script_gen.h:
```diff
@@ -0,0 +1,44 @@
+#include <rapidcheck/gen/Container.h>
+#include <rapidcheck/gen/Select.h>
+
+typedef std::pair<CScript, std::vector<CKey>> SPKCKeyPair;
```
nit: `using`?

Empact commented:
> > why gen::nonEmpty here?
> 
> I believe blocks, by consensus, are required to have at least one transaction in them (citation needed).

I believe that's correct. Relevant check in `CheckBlock` here: https://github.com/bitcoin/bitcoin/blob/4de0b5f39cc35636d499ad29ee3c63384b13fc76/src/validation.cpp#L3112-L3118

fanquake commented:
Rebooted the tests, as they were still stuck on the initial linting error.

This would be a good opportunity to address some of the nits left over from #12775, as well as possibly moving to a newer version of rapidcheck, which we can host at https://github.com/bitcoin-core. 
I bought that up on IRC with @laanwj [here](https://botbot.me/freenode/bitcoin-core-dev/2018-10-20/?msg=105598194&page=2).

We should probably also pull in the depends documentation update from #14171.

FWIW I updated the PR description to point to #12775 instead of ##8469.

practicalswift commented on src/test/gen/block_gen.h:
```diff
@@ -0,0 +1,58 @@
+#include <rapidcheck/Gen.h>
+
+
+typedef std::tuple<int32_t, uint256, uint256, uint32_t, uint32_t, uint32_t> BlockHeaderTup;
```
Nit: Use `using` instead :-)

practicalswift commented on src/test/gen/transaction_gen.h:
```diff
@@ -0,0 +1,124 @@
+#include <rapidcheck/Gen.h>
+#include <rapidcheck/gen/Predicate.h>
+
+typedef std::tuple<const CTxOut, const CTransaction, const int> SpendingInfo;
```
Nit: Use `using` instead :-)

Christewart commented:
So I tried to adjust our build script that is found [here](https://github.com/bitcoin/bitcoin/blob/master/depends/packages/rapidcheck.mk#L11) and failed to sift through the cmake weeds. 

It isn't as trivial as running 

```
$ cmake . 
$ make install 
```

since we needs extra functionality located in this package 

https://github.com/bitcoin-core/rapidcheck/tree/master/extras

specifically the `boost_test` package. This is what is used to hook into the existing boost testing framework. 

It would be much appreciated if someone who is more adept at `cmake` commenting about how to properly add that package in `extras` to have it properly installed in the `rapidcheck.mk` file.

Empact commented:
@Christewart have you seen this? https://github.com/emil-e/rapidcheck/blob/master/doc/boost_test.md

Christewart commented:
@empact Yes! That is how I hook nicely into the existing boost testing framework. 

What I am currently blocked on is adding that submodule to our ci build. Please see the comment above for details.

With these proeprties, if they are enabled, you can just run them simply with 

```
$ ./test_bitcoin 
```

since we integrate with `boost_test`

Christewart commented:
It looks like there is now an easy configuration option in rapidcheck to allow for the installing of `extra/boost_test` by setting a simple configuration option: 

https://github.com/emil-e/rapidcheck/issues/222#issuecomment-443259105

We will need to update the rapidcheck dependency to at least https://github.com/emil-e/rapidcheck/commit/5f0eb302f28979395a070bbdfde1f6d4cad94409

fanquake commented:
@Christewart I've made an attempt to use the new install flags in #14853.

When I merge these changes on top of that PR (using the new rapidcheck) I'm seeing some compilation errors. This is on a macOS 10.14.1 machine, using Xcode 10.1 (10B61):
```
In file included from test/gen/transaction_gen.cpp:1:
In file included from ./test/gen/transaction_gen.h:4:
In file included from ./test/gen/crypto_gen.h:4:
In file included from ./key.h:10:
In file included from ./pubkey.h:10:
In file included from ./hash.h:11:
In file included from ./crypto/sha256.h:10:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string:477:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string_view:176:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__string:56:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:643:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:662:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:227:10: error: static_assert failed due to requirement '__can_bind_reference<std::__1::vector<CTxIn, std::__1::allocator<CTxIn> > >()' "Attempted to construct a reference element in a tuple with an rvalue"
        {static_assert(__can_bind_reference<_Tp>(),
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:411:15: note: in instantiation of function template specialization 'std::__1::__tuple_leaf<1, const std::__1::vector<CTxIn, std::__1::allocator<CTxIn> > &, false>::__tuple_leaf<std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, void>' requested here
            : __tuple_leaf<_Indx, _Tp>(_VSTD::forward<typename tuple_element<_Indx,
              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:828:15: note: in instantiation of function template specialization 'std::__1::__tuple_impl<std::__1::__tuple_indices<0, 1, 2, 3>, int, const std::__1::vector<CTxIn, std::__1::allocator<CTxIn> > &, const std::__1::vector<CTxOut, std::__1::allocator<CTxOut> > &, unsigned int>::__tuple_impl<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, void>' requested here
            : __base_(_VSTD::forward<_Tuple>(__t)) {}
              ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/shrinkable/Transform.hpp:20:37: note: in instantiation of function template specialization 'std::__1::tuple<int, const std::__1::vector<CTxIn, std::__1::allocator<CTxIn> > &, const std::__1::vector<CTxOut, std::__1::allocator<CTxOut> > &, unsigned int>::tuple<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, false>' requested here
  U value() const { return m_mapper(m_shrinkable.value()); }
                                    ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Shrinkable.hpp:26:44: note: in instantiation of member function 'rc::shrinkable::detail::MapShrinkable<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)>::value' requested here
  T value() const override { return m_impl.value(); }
                                           ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Shrinkable.hpp:22:12: note: in instantiation of member function 'rc::Shrinkable<CTransaction>::ShrinkableImpl<rc::shrinkable::detail::MapShrinkable<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)> >::value' requested here
  explicit ShrinkableImpl(Args &&... args)
           ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Shrinkable.hpp:101:27: note: (skipping 3 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  shrinkable.m_impl = new ShrinkableImpl(std::forward<Args>(args)...);
                          ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Gen.hpp:39:12: note: in instantiation of member function 'rc::gen::detail::MapGen<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)>::operator()' requested here
    return m_impl(random, size);
           ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Gen.hpp:34:3: note: in instantiation of member function 'rc::Gen<CTransaction>::GenImpl<rc::gen::detail::MapGen<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)> >::generate' requested here
  GenImpl(Args &&... args)
  ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/Gen.hpp:58:18: note: in instantiation of function template specialization 'rc::Gen<CTransaction>::GenImpl<rc::gen::detail::MapGen<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)> >::GenImpl<rc::gen::detail::MapGen<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)> >' requested here
    : m_impl(new GenImpl<Decay<Impl>>(std::forward<Impl>(impl))) {}
                 ^
/Users/michael/github/bitcoin/depends/x86_64-apple-darwin18.2.0/share/../include/rapidcheck/gen/Transform.hpp:78:10: note: in instantiation of function template specialization 'rc::Gen<CTransaction>::Gen<rc::gen::detail::MapGen<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)>, void>' requested here
  return detail::MapGen<T, Decay<Mapper>>(std::move(gen),
         ^
./test/gen/transaction_gen.h:95:21: note: in instantiation of function template specialization 'rc::gen::map<std::__1::tuple<int, std::__1::vector<CTxIn, std::__1::allocator<CTxIn> >, std::__1::vector<CTxOut, std::__1::allocator<CTxOut> >, unsigned int>, (lambda at ./test/gen/transaction_gen.h:97:13)>' requested here
        return gen::map(gen::tuple(gen::arbitrary<int32_t>(),
```

The entire build output is available [here](https://gist.github.com/fanquake/0b21f286f0dc737295e4e6172274eca0).

fanquake commented:
@Christewart Could you rebase this, now that #14853 has been merged?

Christewart commented:
Rebased this branch, but i am getting a new error when running these tests locally. This doesn't seem related to my PR AFAICT

```
chris@chris:~/dev/bitcoin/src/test$ ./test_bitcoin
Running 365 test cases...
Using configuration: seed=16472686160129483926
Error: Specified -walletdir "/tmp/test_common_Bitcoin Core/1555532449_497584857/tempdir/path_does_not_exist" does not exist
Error: Specified -walletdir "/tmp/test_common_Bitcoin Core/1555532449_512688440/tempdir/not_a_directory.dat" is not a directory
Error: Specified -walletdir "wallets" is a relative path

```
```

practicalswift commented:
@Christewart Yes, those annoying errors are unrelated: see discussion in #15352 :-)

fanquake commented on src/test/gen/crypto_gen.cpp:
```diff
@@ -1,6 +1,3 @@
-// Copyright (c) 2018 The Bitcoin Core developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
 #include <test/gen/crypto_gen.h>
```
Please put this back. Same in the `.h`.

fanquake reviewed (commented):
@Christewart Do you have time to continue with this RapidCheck test work going forward? If so, could you rebase this on master (to pull in a build fix), address outstanding nits and squash your commits etc.

I checked this out and ran the tests. For some reason I thought I had (or used to have?) `rapidcheck` installed via brew, but it doesn't seem to be available from there at the moment.

I'm wondering what the state of the RapidCheck project is, as the [repo](https://github.com/emil-e/rapidcheck) hasn't seen any activity for 6 months, and the author seems to have [gone quiet on GitHub](https://github.com/emil-e). There are a few issues as well as PRs piling up on that repo.

jonatack commented:
@fanquake I wrote @Christewart about this PR 3 days ago and he might have some time this week.

Christewart commented:
So i rebased onto master, and fixed various things that needed to be fixed with API changes in master. 

>Do you have time to continue with this RapidCheck test work going forward? If so, could you rebase this on master (to pull in a build fix), address outstanding nits and squash your commits etc.

>I'm wondering what the state of the RapidCheck project is, as the repo hasn't seen any activity for 6 months, and the author seems to have gone quiet on GitHub. There are a few issues as well as PRs piling up on that repo.

I think this is concerning as well, I do not have the time or know how to maintain rapidcheck itself. I see [you have an issue on the repo](https://github.com/emil-e/rapidcheck/issues/232) about keeping rapidcheck working with newer versions of gcc, and it hasn't been answered which is concerning. 

>I checked this out and ran the tests. For some reason I thought I had (or used to have?) rapidcheck installed via brew, but it doesn't seem to be available from there at the moment.

I don't think rapidcheck can be installed with brew (admittedly, I don't have a mac). I think you need to do 

```bash
cmake . && make install
-- Configuring done
-- Generating done
-- Build files have been written to: /home/chris/dev/rapidcheck
[100%] Built target rapidcheck
Install the project...
-- Install configuration: ""
-- Up-to-date: /usr/local/lib/librapidcheck.a
-- Up-to-date: /usr/local/include
-- Up-to-date: /usr/local/include/rapidcheck.h
-- Up-to-date: /usr/local/include/rapidcheck
....
-- Up-to-date: /usr/local/include
-- Up-to-date: /usr/local/include/rapidcheck
-- Up-to-date: /usr/local/include/rapidcheck/boost_test.h

```

jonatack commented:
@Christewart that worked for me for installing rapidcheck. I'd like to test this soon (and look at the current state of C++ prop-based testing libs).

jonatack commented on src/Makefile.test.include:
```diff
@@ -153,11 +153,25 @@ BITCOIN_TESTS =\
 if ENABLE_PROPERTY_TESTS
 BITCOIN_TESTS += \
-  test/key_properties.cpp
+  test/block_properties.cpp \
```
nit: could update "2013-2016" in this file's copyright header

jonatack reviewed (commented):
Concept ACK, approach ACK. Built and ran all tests successfully, both this branch and also rebased on latest master, with Linux Debian and a fresh build of RapidCheck on master at https://github.com/emil-e/rapidcheck/commit/d9482c683429fe79122e3dcab14c9655874aeb8e.

```
  CXX      test/gen/test_bitcoin-block_gen.o
  CXX      test/gen/test_bitcoin-bloom_gen.o
  CXX      test/gen/test_bitcoin-crypto_gen.o
  CXX      test/gen/test_bitcoin-script_gen.o
  CXX      test/gen/test_bitcoin-transaction_gen.o

[...]

  CXX      test/test_bitcoin-block_properties.o
  CXX      test/test_bitcoin-bloom_properties.o
  CXX      test/test_bitcoin-key_properties.o
  CXX      test/test_bitcoin-merkleblock_properties.o
  CXX      test/test_bitcoin-script_properties.o
  CXX      test/test_bitcoin-transaction_properties.o

[...]

Running tests: block_properties from test/block_properties.cpp
Running tests: bloom_properties from test/bloom_properties.cpp
Running tests: key_properties from test/key_properties.cpp
Running tests: merkleblock_properties from test/merkleblock_properties.cpp
Running tests: script_properties from test/script_properties.cpp
Running tests: transaction_properties from test/transaction_properties.cpp
```

@Christewart could you put this PR into final ACK-ready form? I'm guessing multiple commits for the different tests or Makefile changes would probably be fine (given the work put into it) but would be good to squash the linter, typo, rebase, and clang commits.

Rapidcheck build aspects could be addressed in a separate PR as this one concentrates on adding tests.

Note that Rapidcheck has recently seen a bit of renewed maintainer activity, though afaik https://github.com/emil-e/rapidcheck/issues/232 filed by @fanquake hasn't been addressed yet.

Christewart commented:
@jonatack I think everything is good to go now?

fanquake commented:
Closing this in light of #18514.
